### PR TITLE
Support array `flatten` sql function

### LINKED ISSUE #6995: Implement `flatten` function
### Is your feature request related to a problem or challenge?

## Summary

| Characteristic        | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |
| --------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Function name:        | `flatten`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
| Aliases:              | -                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
| Original function?:   | No                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
| Function Description: | <b>Azure DataBricks:</b> Returns an array of the elements in the union of array1 and array2 without duplicates. </br> <b>Spark SQL:</b> Creates a single array from an array of arrays column. </br> <b>ClickHouse:</b> Converts an array of arrays to a flat array. Function: Applies to any depth of nested arrays. Does not change arrays that are already flat. The flattened array contains all the elements from all source arrays. </br> <b>DuckDB:</b> Concatenate a list of lists into a single list. This only flattens one level of the list (see examples). |
| Sources:              | [Concept](https://github.com/apache/arrow-datafusion/discussions/6855) [Azure](https://learn.microsoft.com/en-us/azure/databricks/sql/language-manual/functions/flatten) [DuckDB](https://duckdb.org/docs/sql/functions/nested) [ClickHouse](https://clickhouse.com/docs/en/sql-reference/functions/array-functions#arrayflatten) [Spark](https://sparkbyexamples.com/spark/spark-sql-array-functions/)                                                                                                                                                           |

## Examples:

```
D select flatten([[1, 2, 3], [4, 5, 6]]);
┌──────────────────────────────────────────────────────────────────────────────┐
│ flatten(main.list_value(main.list_value(1, 2, 3), main.list_value(4, 5, 6))) │
│                                   int32[]                                    │
├──────────────────────────────────────────────────────────────────────────────┤
│ [1, 2, 3, 4, 5, 6]                                                           │
└──────────────────────────────────────────────────────────────────────────────┘
D select flatten([[[1, 1], [2, 2], [3, 3]], [[4, 4], [5, 5], [6, 6]]]);
┌────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ flatten(main.list_value(main.list_value(main.list_value(1, 1), main.list_value(2, 2), main.list_value(3, 3)), main.list_value(main.list_value(4, 4), main.li…  │
│                                                                           int32[][]                                                                            │
├────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ [[1, 1], [2, 2], [3, 3], [4, 4], [5, 5], [6, 6]]                                                                                                               │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```


### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #6995

### CODE REVIEW

**Comment by jayzhan211** [comment]
@alamb @izveigor Ready for review, thanks.

File: datafusion/core/tests/sqllogictests/test_files/array.slt [hunk 1]
```diff
@@ -2319,6 +2314,31 @@ select array_concat(column1, [7]) from arrays_values_v2;
...
+[1, 2, 1, 3, 2] [1, 10, 11] [1, 2, 3, , 4, , 5] [1.1, 2.2, 3.3, 4.4]
+
+query ????
+select column1, column2, column3, column4 from arrays_flatten_table;
+----
+[[1], [2], [3]] [[[1, 2, 3]], [[4, 5]], [[6]]] [[[[1]]], [[[2, 3]]]] [[1.0], [2.1, 2.2], [3.2, 3.3, 3.4]]
+[[1, 2], [3, 4], [5, 6]] [[[8]]] [[[[1, 2]]], [[[3]]]] [[1.0, 2.0], [3.0, 4.0], [5.0, 6.0]]
+
+query ????
+select array_flatten(column1),
```
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/core/tests/sqllogictests/test_files/array.slt:2333]: okay, let me change the name

File: datafusion/physical-expr/src/array_expressions.rs [hunk 2]
```diff
@@ -1738,6 +1738,53 @@ pub fn cardinality(args: &[ArrayRef]) -> Result<ArrayRef> {
     Ok(Arc::new(result) as ArrayRef)
 }
 
+// Create new offsets that are euqiavlent to `flatten` the array.
+fn get_offsets_for_flatten(
+    offsets: OffsetBuffer<i32>,
+    indexes: OffsetBuffer<i32>,
+) -> OffsetBuffer<i32> {
+    let buffer = offsets.into_inner();
+    let offsets: Vec<i32> = indexes.iter().map(|i| buffer[*i as usize]).collect();
```
Comment (alamb) [inline datafusion/physical-expr/src/array_expressions.rs:1747]: This is a very nice implementation 👌

**Review by alamb — APPROVED** [review]
I agree -- this looks great @jayzhan211  thank you

And thank you for the review @izveigor

**Pull request merged**